x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Lookup.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.plan.logical;
+
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamInput;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamOutput;
+import org.elasticsearch.xpack.esql.plan.logical.join.Join;
+import org.elasticsearch.xpack.esql.plan.logical.local.LocalRelation;
+import org.elasticsearch.xpack.ql.capabilities.Resolvables;
+import org.elasticsearch.xpack.ql.expression.Attribute;
+import org.elasticsearch.xpack.ql.expression.Expression;
+import org.elasticsearch.xpack.ql.expression.NamedExpression;
+import org.elasticsearch.xpack.ql.plan.logical.LogicalPlan;
+import org.elasticsearch.xpack.ql.plan.logical.UnaryPlan;
+import org.elasticsearch.xpack.ql.tree.NodeInfo;
+import org.elasticsearch.xpack.ql.tree.Source;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+import static org.elasticsearch.xpack.esql.expression.NamedExpressions.mergeOutputAttributes;
+
+/**
+ * Looks up values from the associated {@code tables}.
+ * The class is used only during the analysis phased, after which it is replaced by a {@code Join}.
+ */
+public class Lookup extends UnaryPlan {
+    private final Expression tableName;
+    private final List<NamedExpression> matchFields;
+    // initialized during the analysis phase for output and validation
+    // afterward, it is converted into a Join (BinaryPlan) hence why here it is not a child
+    private final LocalRelation localRelation;
+    private List<Attribute> lazyOutput;
+
+    public Lookup(
+        Source source,
+        LogicalPlan child,
+        Expression tableName,
+        List<NamedExpression> matchFields,
+        @Nullable LocalRelation localRelation
```

Comment:
This theoretically allows for lookups with multivalued blocks. I wonder if we want to
- Disable this through validation, or
- somehow test this?

We may not want to allow `LOOKUP` where the supplied table has multivalues, but later `INLINESTATS` will force us to consider this if we want to allow `INLINESTATS` with `VALUES`, no?

Reply:
I also realized: what happens when the key column is multivalued? And how's it work when the key is composite - how do we define a match - only when all values coincide, or does `a: [1,2], b:[3,4]` in the index match with `a:1, b:4` in the table?

Reply:
We have a *ton* of tests around multivalued keys and values actually. But, yeah, we should disable multivalued table keys, at least for now. We don't need them for INLINESTATS because that'll always join on the grouping columns which never output multivalued fields.

Now, when the incoming field is multivalued, that gets *fun*. We match the combinatorial explosion of all combinations of the keys. Specifically because that's what INLINESTATS will need - at least, that's what it'll need if we want it to mirror behavior of aggs. Even more fun - there isn't any real ordering in that combinatorial explosion. So, while it's correct, it might not be useful.